source/h2_decoder.c

```diff
@@ -195,52 +201,64 @@ int aws_h2_decode(struct aws_h2_decoder *decoder, struct aws_byte_cursor *data)
     AWS_PRECONDITION(decoder);
     AWS_PRECONDITION(data);
 
+    AWS_FATAL_ASSERT(!decoder->has_errored);
+
+    int err = AWS_OP_SUCCESS;
+
     while (data->len) {
-        const uint64_t bytes_required = decoder->state.bytes_required == WAIT_FOR_FULL_PAYLOAD
-                                            ? decoder->frame_in_progress.payload_len
-                                            : decoder->state.bytes_required;
-        int err = AWS_OP_SUCCESS;
+        const uint32_t bytes_required = decoder->state.bytes_required;
         if (!decoder->scratch.len && data->len >= bytes_required) {
             /* Easy case, there is no scratch and we have enough data, so just send it to the state */
             const char *current_state = decoder->state.name;
-            const uint64_t pre_state_data_len = data->len;
+            const size_t pre_state_data_len = data->len;
 
             err = decoder->state.fn(decoder, data);
+            if (err) {
+                goto handle_error;
+            }
 
-            const uint64_t data_processed = pre_state_data_len - data->len;
+            const size_t data_processed = pre_state_data_len - data->len;
             if (bytes_required > 0 && data_processed != bytes_required) {
                 DECODER_LOGF(
                     DEBUG,
                     decoder,
-                    "Decoder state %s requested %" PRIu64 " bytes of data, but only used %" PRIu64,
+                    "Decoder state %s requested %" PRIu32 " bytes of data, but only used %zu",
                     current_state,
                     bytes_required,
                     data_processed);
             }
         } else {
             /* In every other case, we have to copy to scratch */
-            uint64_t bytes_to_read = bytes_required - decoder->scratch.len;
+            size_t bytes_to_read = bytes_required - decoder->scratch.len;
             bool will_finish_state = true;
-            /* Not enough in this cursor, need to read as much as possible and then come back */
-            if (bytes_to_read > data->len) {
+            if (AWS_UNLIKELY(bytes_required == 0 && decoder->scratch.len)) {
+                /* If this is a streaming state, and there's data in scratch, just send it all in now w/o reading */
+                bytes_to_read = 0;
```

Comment:
Is this possible? If this can only happen when we program our states wrong, let's turn it into an ASSERT